### PR TITLE
feat: `Report False Positive` API implementation [ROAD-543][ROAD-548]

### PR DESCRIPTION
**Pull request by ArtsiomCh**
Tests are omitted due to the fact that implemented functionality is related to UX and Net and is very hard(and very fragile on changes) to implement without numerous mocks -> all my attempts ending up testing mostly that mocks...

image
3 attempts to execute `feedback/sast` call was made:
```
2022-04-28 17:55:33,590 [  68842]  DEBUG - .snyk.plugin.net.SnykApiClient - Executing request to feedback/sast 
2022-04-28 17:55:34,291 [  69543]   WARN - .snyk.plugin.net.SnykApiClient - Failed to execute `feedback/sast` call: {"code":404,"message":"bad API request, please contact support@snyk.io for assistance","error":"unsupported url"} 
2022-04-28 17:55:34,291 [  69543]  DEBUG - .snyk.plugin.net.SnykApiClient - Executing request to feedback/sast 
2022-04-28 17:55:34,826 [  70078]   WARN - .snyk.plugin.net.SnykApiClient - Failed to execute `feedback/sast` call: {"code":404,"message":"bad API request, please contact support@snyk.io for assistance","error":"unsupported url"} 
2022-04-28 17:55:34,826 [  70078]  DEBUG - .snyk.plugin.net.SnykApiClient - Executing request to feedback/sast 
2022-04-28 17:55:35,567 [  70819]   WARN - .snyk.plugin.net.SnykApiClient - Failed to execute `feedback/sast` call: {"code":404,"message":"bad API request, please contact support@snyk.io for assistance","error":"unsupported url"} 
2022-04-28 17:55:35,571 [  70823]   WARN - in.ui.SnykBalloonNotifications - Unable to send report, please contact support@snyk.io for assistance 
```

### CODE REVIEW

File: src/main/kotlin/io/snyk/plugin/services/SnykApiService.kt [hunk 1]
```diff
@@ -77,24 +81,44 @@
...
+            currentUniqSnykApiClient?.disableSslVerification != disableSslVerification
+        ) {
+            log.debug("Creating new SnykApiClient")
+            currentUniqSnykApiClient = try {
+                val retrofit = createRetrofit(token, baseUrl, disableSslVerification)
+                UniqSnykApiClient(
+                    snykApiClient = SnykApiClient(retrofit),
+                    token = token,
+                    baseUrl = baseUrl,
+                    disableSslVerification = disableSslVerification
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/services/SnykApiService.kt:106]: Please add the parameter that we could enable request logging if needed

File: src/main/kotlin/io/snyk/plugin/ui/toolwindow/SuggestionDescriptionPanel.kt [hunk 2]
```diff
@@ -399,12 +402,40 @@
~                    addActionListener { e ->
~                        val dialog = ReportFalsePositiveDialog(
~                            project,
                             titlePanel(insets = JBUI.insetsBottom(10), indent = 0)
                         )
                         if (dialog.showAndGet()) {
-                            this.isEnabled = false
-                            this.text = FALSE_POSITIVE_REPORTED_TEXT
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/toolwindow/SuggestionDescriptionPanel.kt:406]: why fixme?

File: src/main/kotlin/io/snyk/plugin/ui/toolwindow/SuggestionDescriptionPanel.kt [hunk 3]
```diff
@@ -399,12 +402,40 @@
...
+                                    userPublicId = snykApiService.userId ?: "",
+                                    startLine = suggestionRange?.startRow ?: 1,
+                                    endLine = suggestionRange?.endRow ?: 1,
+                                    primaryFilePath = snykCodeFile.virtualFile.path,
+                                    vulnName = suggestion.rule,
+                                    fileContents = "dialog.result" //fixme
+                                )
+                            )
+                            if (snykApiService.reportFalsePositive(payload)) {
+                                this.isEnabled = false
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/toolwindow/SuggestionDescriptionPanel.kt:424]: Would now be the time to update it?

File: src/main/kotlin/io/snyk/plugin/ui/toolwindow/SuggestionDescriptionPanel.kt [hunk 4]
```diff
@@ -399,12 +402,40 @@
...
+                                this.text = FALSE_POSITIVE_REPORTED_TEXT
+                                SnykBalloonNotificationHelper.showInfo(
+                                    // todo: correct text with Andy
+                                    "False Positive reported. Thank you!",
+                                    project
+                                )
+                                // todo: disable re-report per session/project/application/token/org ?
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/toolwindow/SuggestionDescriptionPanel.kt:431]: see above - please finalize text

File: src/main/kotlin/io/snyk/plugin/ui/toolwindow/SuggestionDescriptionPanel.kt (hunk 2 above)
Comment (ArtsiomCh) [reply src/main/kotlin/io/snyk/plugin/ui/toolwindow/SuggestionDescriptionPanel.kt:406]: to not forget to change after rebase :)

File: src/main/kotlin/io/snyk/plugin/ui/toolwindow/SuggestionDescriptionPanel.kt (hunk 3 above)
Comment (ArtsiomCh) [reply src/main/kotlin/io/snyk/plugin/ui/toolwindow/SuggestionDescriptionPanel.kt:423]: Will discuss with Andy. Imho the best way to inform user about succeed post - will be disable button with appropriate text. At least that how Ide do it (in Exception reporter for example)

**Comment by bastiandoetsch** [comment]
> Tests are omitted due to the fact that implemented functionality is related to UX and Net and is very hard(and very fragile on changes) to implement without numerous mocks -> all my attempts ending up testing mostly that mocks...

Got some questions: 
- Why can't you test the API with the mock webserver from retrofit? Or just mock the ApiClient with mockk in Tests?
-  Regarding the UI changes, I'll have a look in the code and comment there where a test would make sense...

File: src/main/kotlin/io/snyk/plugin/net/SnykApiClient.kt [hunk 5]
```diff
@@ -8,42 +8,36 @@ import retrofit2.Retrofit
  * Main entrypoint for using Snyk API.
  */
 class SnykApiClient constructor(
-    private val retrofit: Retrofit
+    retrofit: Retrofit
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/net/SnykApiClient.kt:11]: with a retrofit mock, testing the ApiClient should be possible & straightforward

File: src/main/kotlin/io/snyk/plugin/services/SnykApiService.kt [hunk 6]
```diff
@@ -24,11 +26,14 @@ class SnykApiService : Disposable {
...
         get() {
-            return snykApiClient?.sastSettings(pluginSettings().organization)
+            return getSnykApiClient()?.sastSettings(pluginSettings().organization)
         }
 
     val userId: String?
-        get() = snykApiClient?.userId
+        get() = getSnykApiClient()?.getUserId()
+
+    fun reportFalsePositive(payload: FalsePositivePayload): Boolean =
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/services/SnykApiService.kt:35]: If you inject a SnykApiClient as a mock, it should be possible to test expected service behaviour without doing the actual network call.

**Review by bastiandoetsch — COMMENTED** [review]
Yeah, I agree that testing the UI stuff would not really help. But the `SnykApiService` and `SnykApiClient` can and should be tested.

File: src/main/kotlin/io/snyk/plugin/net/SnykApiClient.kt (hunk 5 above)
Comment (ArtsiomCh) [reply src/main/kotlin/io/snyk/plugin/net/SnykApiClient.kt:11]: done

File: src/main/kotlin/io/snyk/plugin/services/SnykApiService.kt (hunk 6 above)
Comment (ArtsiomCh) [reply src/main/kotlin/io/snyk/plugin/services/SnykApiService.kt:35]: here I skip tests for SnykApiService as it just proxying call to `SnykApiClient.reportFalsePositive(payload)` and the latter tested in SnykApiClientTest

**Review by bastiandoetsch — APPROVED** [review]
LGTM, thanks for adding the test :)

**Pull request merged**